kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/EncryptionDekCache.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.filter.encryption.inband;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.Executor;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.benmanes.caffeine.cache.AsyncLoadingCache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+
+import io.kroxylicious.filter.encryption.EncryptionScheme;
+import io.kroxylicious.filter.encryption.dek.CipherSpec;
+import io.kroxylicious.filter.encryption.dek.Dek;
+import io.kroxylicious.filter.encryption.dek.DekManager;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+public class EncryptionDekCache<K, E> {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(EncryptionDekCache.class);
+
+    public static final int NO_MAX_CACHE_SIZE = -1;
+
+    private record CacheKey<K>(K kek, CipherSpec cipherSpec) {}
+
+    private static <K> CacheKey<K> cacheKey(EncryptionScheme<K> encryptionScheme) {
+        return new CacheKey<>(encryptionScheme.kekId(), CipherSpec.AES_128_GCM_128);
+    }
+
+    private final DekManager<K, E> dekManager;
+
+    private final AsyncLoadingCache<CacheKey<K>, Dek<E>> dekCache;
+
+    public EncryptionDekCache(
+                              @NonNull DekManager<K, E> dekManager,
+                              @Nullable Executor dekCacheExecutor,
+                              int dekCacheMaxItems) {
+        this.dekManager = dekManager;
+        Caffeine<Object, Object> cache = Caffeine.newBuilder();
+        if (dekCacheMaxItems != NO_MAX_CACHE_SIZE) {
+            cache = cache.maximumSize(dekCacheMaxItems);
+        }
+        if (dekCacheExecutor != null) {
+            cache = cache.executor(dekCacheExecutor);
+        }
+
+        this.dekCache = cache
+                .removalListener(this::afterCacheEviction)
+                .buildAsync(this::requestGenerateDek);
+    }
+
+    /** Invoked by Caffeine when a DEK needs to be loaded */
+    private CompletableFuture<Dek<E>> requestGenerateDek(@NonNull CacheKey<K> cacheKey,
+                                                         @NonNull Executor executor) {
+        return dekManager.generateDek(cacheKey.kek(), cacheKey.cipherSpec())
+                .thenApply(dek -> {
+                    if (LOGGER.isTraceEnabled()) {
+                        LOGGER.trace("Adding DEK to cache: {}", dek);
+                    }
+                    dek.destroyForDecrypt();
```

Comment:
Should put a comment on this. I assume this is so that when it expires from the encrypt cache the DEK will be destroyed and this code will have to change once we start eagerly populating the decrypt cache at the same time.

What do you think of moving that into the `DekManager` api with methods like `generateEncryptDecryptDek` and `generateDecryptDek`?

Reply:
You're correct that this will change once we're sharing DEKs between the two paths. Here's how I thought that might work (which might not work out in reality):

* The `InBandEncryptionManager` would have a reference to the `InBandDecryptionManager`
* The `InBandDecryptionManager` would have a method `maybeCacheDek()` which the `InBandEncryptionManager` would call when it gets a new DEK
* The `maybeCacheDek()` method would have the responsibility of adding the new DEK to the decryption cache, or not.
* Right now we can easily support `always` adding the DEK to the decryption cache, or `never` doing so. In the future we might want to explore adding it when there's an existing fetch session (`when-session-exists`). Depending on what we find out the performance implications to be this is potentially something which could be configurable. Or not if we find a one-size-fits-all approach is acceptable.
* When `maybeCacheDek()` doesn't add the DEK to the cache it would just immediately call `closeForDecrypt. When the DEK was added to the cache `closeForDecrypt` would get called on cache eviction.